### PR TITLE
[ci] Fixing the doc generation jobs

### LINKED ISSUE #1074: CI docs all job is broken
Some issue with the temporary directory being janked underneath the last jobs. Perhaps there is a size/node limit per directory?

```
Generating documentation for stm32l486zgt7 ...
Executing: (cd stm32l486zgt7/modm/docs/ && doxypress doxypress.json)
Error 2 generating documentation for device stm32l486zgt7.
Generating documentation for stm32f030f4p6-demo ...
Executing: (cd stm32f030f4p6-demo/modm/docs/ && doxypress doxypress.json)
Error 2 generating documentation for device stm32f030f4p6-demo.
Generating documentation for stm32l432kcu6 ...
Executing: (cd stm32l432kcu6/modm/docs/ && doxypress doxypress.json)
Error 2 generating documentation for device stm32l432kcu6.
You need to provide the programmer via a custom OpenOCD config!

# Black Pill

STM32F103C8T6 "Black Pill" Minimum System Development Board

Cheap and bread-board-friendly board for STM32 F1 series.
Sold for less than 2 USD on well known Internet shops from China.

https://stm32-base.org/boards/STM32F103C8T6-Black-Pill.html

## Programming

Since the board doesn't have a programmer on-board, you need to use your
own and *specify* which one you're using in a custom `openocd.cfg` file:

```
# Replace this with your custom programmer
source [find interface/stlink-v2.cfg]

# If you use the clone CKS32F103C8T6 chip you need to overwrite this ID
#set CPUTAPID 0x2ba01477

# To select a specific programmer you can specify its serial number
#hla_serial "\x53\x3f\x6f\x06\x50\x77\x50\x57\x12\x17\x14\x3f"
# You can discover the serial via `stlink --hla-serial` or `st-info --hla-serial`.
```

Then include this file in your build options like so:

```xml
<library>
  <extends>modm:black-pill</extends>
  <options>
    <option name="modm:build:openocd.cfg">openocd.cfg</option>
  </options>
</library>
```

TinyUSB in modm does not currently use the FreeRTOS abstraction layer and is not thread-safe with FreeRTOS threads.
TinyUSB in modm does not currently use the FreeRTOS abstraction layer and is not thread-safe with FreeRTOS threads.
Generating documentation for black-pill-f103 ...
Executing: (cd black-pill-f103/modm/docs/ && doxypress doxypress.json)
Finished generating documentation for device black-pill-f103.
Deduplicating files for stm32f103c8t6...

Moving black-pill-f103/modm/docs/html -> output/develop/api/black-pill-f103
Error generating documentation for device black-pill-f103: [Errno 2] No such file or directory: '/tmp/tmp4kvu_hht/black-pill-f103/modm/docs/html' -> '/tmp/tmp4kvu_hht/output/develop/api/black-pill-f103'
TinyUSB in modm does not currently use the FreeRTOS abstraction layer and is not thread-safe with FreeRTOS threads.
Generating documentation for stm32g0c1vet6 ...
Executing: (cd stm32g0c1vet6/modm/docs/ && doxypress doxypress.json)
Finished generating documentation for device stm32g0c1vet6.
Deduplicating files for stm32g0c1vet6...

Moving stm32g0c1vet6/modm/docs/html -> output/develop/api/stm32g0c1vet6
Error generating documentation for device stm32g0c1vet6: [Errno 2] No such file or directory: '/tmp/tmp4kvu_hht/stm32g0c1vet6/modm/docs/html' -> '/tmp/tmp4kvu_hht/output/develop/api/stm32g0c1vet6'
>> modm: Recomputing device cache...
Modm Path: /__w/modm/modm
Temporary directory: /tmp/tmp4kvu_hht
Starting to generate documentation...
... for 259 devices, estimated memory footprint is 15370 MB
Traceback (most recent call last):
  File "tools/scripts/docs_modm_io_generator.py", line 151, in main
    results = pool.map(lambda d: subprocess.run(d, shell=True).returncode, list(set(devices)))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 765, in get
    self.wait(timeout)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 762, in wait
    self._event.wait(timeout)
  File "/usr/lib/python3.8/threading.py", line 558, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib/python3.8/threading.py", line 302, in wait
    waiter.acquire()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tools/scripts/docs_modm_io_generator.py", line 272, in <module>
    exit(0 if main() else -1)
  File "tools/scripts/docs_modm_io_generator.py", line 170, in main
    return results.count(0) == len(results)
  File "/usr/lib/python3.8/tempfile.py", line 966, in __exit__
    self.cleanup()
  File "/usr/lib/python3.8/tempfile.py", line 970, in cleanup
    self._rmtree(self.name)
  File "/usr/lib/python3.8/tempfile.py", line 952, in _rmtree
    _rmtree(name, onerror=onerror)
  File "/usr/lib/python3.8/shutil.py", line 722, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/usr/lib/python3.8/shutil.py", line 720, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/tmp/tmp4kvu_hht'
TinyUSB in modm does not currently use the FreeRTOS abstraction layer and is not thread-safe with FreeRTOS threads.
```

### PR DESCRIPTION
**Pull request by salkinium**
Fixes #1074 (hopefully).

### CODE REVIEW

**Comment by salkinium** [comment]
ok, so plain stupid container/runner update doesn't fix it, but seems to make it fail silently. That's… great.

**Comment by salkinium** [comment]
Still failing silently. Perhaps overriding the Python installation with the latest or previous version will fix it?

**Comment by rleh** [comment]
I'll take a (deeper) look later this week.

**Comment by salkinium** [comment]
> You are running out of disk space. The runner will stop working when the machine runs out of disk space. Free space left: 96 MB

Oh, it has nothing to do with the code. We're simply out of disk space. The last job probably did better at symlinking to reduce duplication: https://github.com/modm-io/modm/actions/runs/6318898186

Ok, so we need to split the job up into multiple parts and upload them separately. I'm not going to do that for the 2023q3 release though, since the docs are resonably up-to-date (27. Sept).

**Comment by salkinium** [comment]
Ok, so I have implemented an "incremental compression" scheme, where each generated target inside `develop/api/` is compressed into its own `.tar.gz` folder and then the uncompressed files are deleted. This should prevent us from running out of disk space. Unfortunately, this means that @rleh needs to adapt the hook on the hosting server to also unpack those files.

**Comment by salkinium** [comment]
Well… it worked! I successfully killed the docs website with my compressed targets! There's still some issue with the task return value, but it doesn't seem like we're running out of disk space anymore. Plus it's a little faster now.

**Comment by salkinium** [comment]
Ah, it failed only because of the lack of fibers on the RPi. ARM64 fibers are not implemented.
```
Generating documentation for raspberrypi...
Error generating documentation for device raspberrypi: Cannot resolve dependency ':processing:fiber' of Module(modm:driver:bmi088)!
In file '../../__w/modm/modm/src/modm/driver/inertial/bmi088.lb:22' in function 'prepare':
def prepare(module, options):
```

**Comment by salkinium** [comment]
Ok, cleaned up the code a little more and will merge once the CI passes (except for the RPi fiber issue).

File: tools/scripts/docs_modm_io_generator.py [hunk 1]
```diff
@@ -136,56 +138,60 @@ def main():
...
+                file.unlink()
         if args.compress:
             print("Zipping docs ...")
-            # Zipping may take more than 10 minutes
-            os.system(f"(cd {str(output_dir)} && {'g' if is_running_on_macos else ''}tar --checkpoint=.100 -czf {str(cwd / 'modm-api-docs.tar.gz')} .)")
-            # shutil.make_archive(str(cwd / 'modm-api-docs'), 'gztar', str(output_dir))
+            # Zipping may take some time
+            tar_cmd = "gtar" if is_running_on_macos else "tar"
+            zip_cmd = f'(cd {str(output_dir)} && {tar_cmd} --checkpoint=.100 -czf {str(cwd)}/modm-api-docs.tar.gz .)'
+            subprocess.call(zip_cmd, shell=True)
```
Comment (rleh) [inline tools/scripts/docs_modm_io_generator.py]:
```suggestion
import tarfile

tar = tarfile.open(f"{str(cwd)}/modm-api-docs.tar", "w")
tar.add(str(cwd)}/modm-api-docs.tar.gz)
tar.close()
```

Comment (salkinium) [reply tools/scripts/docs_modm_io_generator.py]: Implemented it, will report back how large the `.tar` file is that the CI uploads.

Comment (salkinium) [reply tools/scripts/docs_modm_io_generator.py]: 1473MB with `.tar` vs 331MB with `.tar.gz`. The upload also does not seem to have been properly deployed on the docs server. So, we're keeping it as is.

**Pull request merged**